drivers/gpio/gpio_shell.c

```diff
@@ -1,187 +1,441 @@
+		} else if (input[i] != line_name[i]) {
+			return (input[i] > line_name[i]) ? 1 : -1;
+		} else if (line_name[i] == '\0') {
+			return 0;
+		}
+		i++;
+	}
+}
+
+static int get_gpio_pin(const struct shell *sh, const struct gpio_ctrl *ctrl, char *line_name)
+{
+	gpio_pin_t pin = PIN_NOT_FOUND;
+	gpio_pin_t i;
+	int result;
+
+	for (i = 0; i < ctrl->ngpios; i++) {
+		result = line_cmp(line_name, ctrl->line_names[i]);
+		if (result == 0) {
+			if ((BIT64(i) & ctrl->reserved_mask) != 0) {
+				shell_error(sh, "Reserved pin");
+				return -EACCES;
+			} else if (pin == PIN_NOT_FOUND) {
+				pin = i;
+			} else {
+				shell_error(sh, "Line name ambiguous");
+				return -EFAULT;
+			}
+		}
+	}
+
+	if (pin == PIN_NOT_FOUND) {
+		shell_error(sh, "Line name not found: '%s'", line_name);
+		return -ENOENT;
+	}
+
+	return pin;
+}
+
+static int get_sh_gpio(const struct shell *sh, char **argv, struct sh_gpio *gpio)
+{
+	const struct gpio_ctrl *ctrl;
+	int ret = 0;
+	int pin;
+
+	ctrl = get_gpio_ctrl(argv[ARGV_DEV]);
+	if (ctrl == NULL) {
+		shell_error(sh, "unknown gpio controller: %s", argv[ARGV_DEV]);
 		return -EINVAL;
 	}
+	gpio->dev = ctrl->dev;
+	pin = shell_strtoul(argv[ARGV_PIN], 0, &ret);
+	if (ret != 0) {
+		pin = get_gpio_pin(sh, ctrl, argv[ARGV_PIN]);
+		if (pin < 0) {
+			return pin;
+		}
+	} else if ((BIT64(pin) & ctrl->reserved_mask) != 0) {
+		shell_error(sh, "Reserved pin");
+		return -EACCES;
+	}
+	gpio->pin = pin;
+
+	return 0;
+}
+
+static int cmd_gpio_conf(const struct shell *sh, size_t argc, char **argv, void *data)
+{
+	gpio_flags_t flags = 0;
+	struct sh_gpio gpio;
+	int ret = 0;
+
+	ret = get_sh_gpio(sh, argv, &gpio);
+	if (ret != 0) {
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
 
-	dev = device_get_binding(argv[args_indx.port]);
+	for (int i = 0; i < strlen(argv[ARGV_CONF]); i++) {
+		switch (argv[ARGV_CONF][i]) {
+		case 'i':
+			flags |= GPIO_INPUT;
+			break;
+		case 'o':
+			flags |= GPIO_OUTPUT;
+			break;
+		case 'u':
+			flags |= GPIO_PULL_UP;
+			break;
+		case 'd':
+			flags |= GPIO_PULL_DOWN;
+			break;
+		case 'h':
+			flags |= GPIO_ACTIVE_HIGH;
+			break;
+		case 'l':
+			flags |= GPIO_ACTIVE_LOW;
+			break;
+		case '0':
+			flags |= GPIO_OUTPUT_INIT_LOGICAL | GPIO_OUTPUT_INIT_LOW;
+			break;
+		case '1':
+			flags |= GPIO_OUTPUT_INIT_LOGICAL | GPIO_OUTPUT_INIT_HIGH;
+			break;
+		default:
+			shell_error(sh, "Unknown: '%c'", argv[ARGV_CONF][i]);
+			shell_help(sh);
+			return SHELL_CMD_HELP_PRINTED;
+		}
+	}
 
-	if (dev != NULL) {
-		index = (uint8_t)atoi(argv[2]);
-		shell_print(sh, "Reading %s pin %d",
-			    argv[args_indx.port], index);
-		rc = gpio_pin_get(dev, index);
-		if (rc >= 0) {
-			shell_print(sh, "Value %d", rc);
-		} else {
-			shell_error(sh, "Error %d reading value", rc);
-			return -EIO;
+	if (((flags & GPIO_INPUT) != 0) == ((flags & GPIO_OUTPUT) != 0)) {
+		shell_error(sh, "must be either input or output");
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
```

Comment:
Non-blocking comment - This duplicates the logic from `gpio_pin_configure()`, which uses asserts.
As a follow up CL, you could create a helper function for this validation that is used both by the shell command and `gpio_pin_configure()`.